### ISSUE TITLE
joining dataframes using a range

### ISSUE THREAD

**Issue by kaizadp** [kaizadp/nmrrr#4]
@bpbond 

I need to bin my spectra based on a set of compound classes. The spectra datafile has a `ppm` column that goes from 0 to ~15, whereas the bins represent only small chunks across this overall range. The code I have assigns the bins, but only subsets the parts that were a match. So I have blank spots for regions that were not a match. How can I assign these bins while also keeping all the non-matches?

Essentially, a `full_join`, but where one of the data-frames has only ranges.

<img width="228" alt="image" src="https://user-images.githubusercontent.com/50244730/154198185-0b948d29-c54d-4c3e-be2a-5bddd47eb4b0.png">

https://github.com/bpbond/nmrrr/blob/f5e4ae4488028ae03795b686f128a735bb168bef/R/1-functions_processing.R#L167-L175

**Comment by bpbond** [comment]
R has a function to do exactly this: `cut`. Let's say we have 
```
dat <- data.frame(x = 0:10)
bins_dat <- data.frame(group=c("aliphatic1", "aliphatic2", NA, "oalkyl", "alphah", NA, "aromatic", "amide", NA), start = c(0.3, 1.3, 2.2, 2.9, 4.1, 4.8, 6.2, 7.8, 8.4))
```

In other words, `bins_dat` looks like this: 
group | start
-- | --
aliphatic1 | 0.3
aliphatic2 | 1.3
NA | 2.2
oalkyl | 2.9
alphah | 4.1
NA | 4.8
aromatic | 6.2
amide | 7.8
NA | 8.4

 Now, with a single `cut` call we can assign the labels we want:
```r
dat$group <- cut(dat$x, bins_dat$start, labels = head(bins_dat$group, -1), right = FALSE)
```
x | group
-- | --
0 | NA
1 | aliphatic1
2 | aliphatic2
3 | oalkyl
4 | oalkyl
5 | NA
6 | NA
7 | aromatic
8 | amide
9 | NA
10 | NA

This is a faster and cleaner approach than using a `join`, I think.

**Comment by kaizadp** [comment]
That would work, but the current `bins_dat` data frame (start-stop columns format) comes from existing input files ([example: Clemente2012.txt](https://github.com/bpbond/nmrrr/blob/master/bins/Clemente2012.txt)). This is in a format that's understandable to NMR users.

So we would need a script to make it long-form, and then identify the gaps (e.g. 2.2-2.9, 4.8-6.2, etc.) and assign those a group of "NA".
^ This would need to be automated, so we can use it on different binsets.

What do you think is the best way to do this?

**Comment by bpbond** [comment]
That does seem easiest to me, but am open to other approaches! So the idea would be:
1. Read in a `group`-`start`-`stop` file, e.g. the one you link above, and fill in any gaps (code with NA for group)
2. Perform the `cut`

Is that correct?

**Comment by kaizadp** [comment]
Yep!

**Comment by bpbond** [comment]
```r

bins <- readr::read_tsv("~/Desktop/Clemente2012.txt")

# identify gaps between bins
gaps <- c(head(bins$stop, -1) != bins$start[-1], FALSE)
# create new gap bins
gapbins <- tibble(group = NA_character_, start = bins$stop[gaps])

newbins <- rbind(bins[c("group", "start")], gapbins)
```

group | start
-- | --
aliphatic1 | 0.3
aliphatic2 | 1.3
oalkyl | 2.9
alphah | 4.1
aromatic | 6.2
amide | 7.8
NA | 2.2
NA | 4.8

I don't know if `cut` requires that the bins be ordered, so that might be one additional step needed.

**Comment by bpbond** [comment]
Sorry, slight modification, I realize we need there to be a final bin. So change the second line above to
```r
gaps <- c(head(bins$stop, -1) != bins$start[-1], TRUE)
```
producing


group | start
-- | --
aliphatic1 | 0.3
aliphatic2 | 1.3
oalkyl | 2.9
alphah | 4.1
aromatic | 6.2
amide | 7.8
NA | 2.2
NA | 4.8
NA | 8.4

**Comment by kaizadp** [comment]
Thanks! I’ll try this.

**Comment by kaizadp** [comment]
The good news - this worked!

The not-so-good news - because there are NAs in the `group` column, it codes the final version as a weird form of "NA" - not recognized as the typical NA, nor as character "NA". So I am unable to filter these out. See screenshots.

![image](https://user-images.githubusercontent.com/50244730/182940026-57b87a4c-499b-474b-83d2-48ace743dda7.png)

![image](https://user-images.githubusercontent.com/50244730/182939852-b08e307c-65b8-4944-b86e-e5cb740169d4.png)


My current workaround is to force-code the NAs as "NANA" in `newbins`, so I can filter it out in the next step.

https://github.com/bpbond/nmrrr/blob/573df028c31643bade70c676f86e62b20cc4eb8e/R/1-functions_processing.R#L179-L194

Do you have a more elegant solution for this?

**Comment by bpbond** [comment]
In the screenshots above, where is "NANA"? I'm confused as to the problem.

**Comment by kaizadp** [comment]
Sorry for the confusion.

The screenshots are when I don't use line 189 ("NANA"). They are showing the problem - unable to filter out the "NA" values. I just used "NANA" because I couldn't understand how to remove the "NA" values.


If you comment out line 189 and then run this code below, you'll get the screenshot. `scratchpad.R` is where I am testing the functions to see how they work.

https://github.com/bpbond/nmrrr/blob/573df028c31643bade70c676f86e62b20cc4eb8e/scratchpad.R#L27-L31

**Comment by bpbond** [comment]
I think this issue has been obviated by #21 and so can be closed?

**Comment by kaizadp** [comment]
Yes!
Closes #22
